psi4/driver/procrouting/diis.py

```diff
@@ -85,83 +116,95 @@ def copier(self, x, new_name):
+            full_name = self.get_name(name, entry_num, item_num)
+            psio = core.IO.shared_object()
+            if len(entry_dims) == 2:
+                quantity = core.Matrix(full_name, *entry_dims)
+                quantity.load(psio, psif.PSIF_LIBDIIS, core.SaveType.SubBlocks)
+            elif len(entry_dims) == 1:
+                quantity = core.Vector(full_name, *entry_dims)
+                quantity.load(psio, psif.PSIF_LIBDIIS)
+        else:
+            raise Exception(f"StoragePolicy {self.storage_policy} not recognized. This is a bug: contact developers.")
+
+        return quantity
+
+
+    def get_dot_product(self, i: int, j: int):
+        """ Get a DIIS dot product. i and j represent entry numbers. """
         key = frozenset([i, j])
         try:
             return self.cached_dot_products[key]
         except KeyError:
-            if self.storage_policy == StoragePolicy.InCore:
-                Ri = self.stored_vectors[i][0]
-                Rj = self.stored_vectors[j][0]
-                dot_product = sum(Rix.vector_dot(Rjx) for Rix, Rjx in zip(Ri, Rj))
-            elif self.storage_policy == StoragePolicy.OnDisk:
-                dot_product = 0
-                psio = core.IO.shared_object()
-                for x, entry_dims in enumerate(self.R_template):
-                    if len(entry_dims) == 2:
-                        Rix = core.Matrix(self.get_name("R", i, x), *entry_dims)
-                        Rjx = core.Matrix(self.get_name("R", j, x), *entry_dims)
-                        Rix.load(psio, psif.PSIF_LIBDIIS, core.SaveType.SubBlocks)
-                        Rjx.load(psio, psif.PSIF_LIBDIIS, core.SaveType.SubBlocks)
-                    elif len(entry_dims) == 1:
-                        Rix = core.Vector(self.get_name("R", i, x), *entry_dims)
-                        Rjx = core.Vector(self.get_name("R", j, x), *entry_dims)
-                        Rix.load(psio, psif.PSIF_LIBDIIS)
-                        Rjx.load(psio, psif.PSIF_LIBDIIS)
-                    else:
-                        raise Exception("R_template may only have 1 or 2 dimensions. This is a bug: contact developers.")
-                    dot_product += Rix.vector_dot(Rjx)
-            else:
-                raise Exception(f"StoragePolicy {self.storage_policy} not recognized. This is a bug: contact developers.")
+            dot_product = 0
+            for item_num in range(len(self.template["gradient"])):
+                Rix = self.load_quantity("gradient", i, item_num)
+                Rjx = self.load_quantity("gradient", j, item_num)
+                dot_product += Rix.vector_dot(Rjx)
 
             self.cached_dot_products[key] = dot_product
             return dot_product
 
 
     def set_error_vector_size(self, *args):
-        template_helper(self.R_template, *args)
+        """ Set the template for the DIIS error. Kept mainly for backwards compatibility. """
+        self.template["gradient"] = template_helper(*args)
```

Comment:
Renamed to "error" rather than "error vector". Psi's current I/O system ilimits the length of names for files you write to disk. This key becomes part of a name later, so I'm saving letters.